Argparser auto-completion and colored help

jnoortheen opened the pull request:
<!--- Thanks for opening a PR on xonsh! Please include a news entry with your PR
to help keep our changelog up to date! There are instructions available here:
https://xon.sh/devguide.html#changelog -->

<!--- If there is specific issue / feature request that this PR is addressing,
please link to the corresponding issue by using the `#issuenumber` syntax.
Thanks again! -->

## TODO:
- [x] update completer to handle options and nargs
- [x] add coloring 

## sub PRs:

I think it is better to split this PR
- [x] support callable objects - https://github.com/xonsh/xonsh/pull/4379
- [x] argparser colored help and auto-completions - https://github.com/xonsh/xonsh/pull/4391
- [x] dirstack - https://github.com/xonsh/xonsh/pull/4446
- [x] xonfig - https://github.com/xonsh/xonsh/pull/4436/
- [x] xontrib - https://github.com/xonsh/xonsh/pull/4441
- [ ] vox - https://github.com/xonsh/xonsh/pull/4437
- [x] run-tests.xsh - https://github.com/xonsh/xonsh/pull/4434


## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

jnoortheen commented:
added motivation here https://github.com/xonsh/xonsh/issues/4284

daniel-shimon commented:
Hey @jnoortheen - can you explain a bit which improvements this PR does? It's a bit hard to get from the code

jnoortheen commented:
It generates `parser.add_argument` parameters from the function signature. I was thinking of using Annotated to pass the metadata that will make it more clear. Is it Ok to add https://pypi.org/project/typing-extensions/

daniel-shimon commented:
AFAIK all of xonsh's requirements need to be optional
So maybe it can be used optionally with fallbacking to Any

jnoortheen commented:
> AFAIK all of xonsh's requirements need to be optional
> So maybe it can be used optionally with fallbacking to Any

Yes I was also hesitant adding another dependency. But this comes from Python core repo. Also I dont find any other way to pass the metadata that passes type-checker.

daniel-shimon commented:
Since the type checker only runs in the CI, you can use this package and add it to the dev requirements file with a fallback to Any on import error. We don't really need the annotations in production anyway

jnoortheen commented:
> Since the type checker only runs in the CI, you can use this package and add it to the dev requirements file with a fallback to Any on import error. We don't really need the annotations in production anyway

Ok that is better.

jnoortheen commented:
after coloring it looks like this 
![image](https://user-images.githubusercontent.com/6702219/119956917-35f77280-bfbf-11eb-83d6-6461ab1669ac.png)

jnoortheen commented:
I have converted some existing aliases to use this class. So that we can have Python API to these functions as well as coloring/completion.

jnoortheen commented:
it is better to merge #4294 and #4270 before this

jnoortheen commented:
@daniel-shimon I have created separate PRs from this one. This one can be reviewed now.

jnoortheen commented:
- added new utility classes xonsh.cli_utils.ArgParserAlias, xonsh.cli_utils.ArgCompleter. These are helper classes, that add coloring and auto-completion support to the alias-commands.
- added auto-completion support for pushd, vox
- improved auto-completion of xonfig, xontrib

jnoortheen commented:
@gforsyth any comment is appreciated.

gforsyth commented:
@jnoortheen -- I'll try to find some time to review this.  I'm fine with the improvements and general functionality -- this PR is BIG, though, and there lots of small stylistic improvements which makes reviewing it hard.

jnoortheen commented:
@gforsyth sure take your time. Please go through each commit. I've split them in the order.

daniel-shimon commented on xonsh/completers/tools.py:
```diff
@@ -201,3 +201,84 @@ def apply_lprefix(comps, lprefix):
+
+        if isinstance(act.choices, dict):  # sub-parsers
+            for choice, sub_parser in act.choices.items():
+                yield RichCompletion(choice, description=sub_parser.description or "")
```
Also passing `description=None` is fine here so no need for the `or ""` part (also relevant to the end of this function)

jnoortheen replied:
the Richcompletion has str annotation and not optional right?

daniel-shimon commented on xonsh/completers/tools.py:
```diff
@@ -201,3 +201,84 @@ def apply_lprefix(comps, lprefix):
+            if current_flag.startswith(flag):
+                yield from _complete_argparser_action(act, **kwargs)
+            elif not current_flag:  # show flags only for root parser
+                yield RichCompletion(flag, description=act.help or "")
```
Passing `description=None` is fine here so no need for the `or ""` part (also relevant a few lines above)

jnoortheen replied:
as stated above description is expected as string as in the annotation.

daniel-shimon commented on xonsh/cli_utils.py:
```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
+
+
+def add_args(parser: ap.ArgumentParser, func: tp.Callable, allowed_params=None) -> None:
+    """From Arg -> parser.add_agument(*a, *kw)"""
```
typo

jnoortheen replied:
just explained that add_args is does what add_argument doing using `Arg` functions result. I think it doesn't do its job :)

daniel-shimon replied:
Yeah yeah the doc is good just a small typo, `agument` instead of `argument`

jnoortheen replied:
oh just pushed changes for that to #4391

daniel-shimon commented on xonsh/cli_utils.py:
```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
+
+def add_args(parser: ap.ArgumentParser, func: tp.Callable, allowed_params=None) -> None:
+    """From Arg -> parser.add_agument(*a, *kw)"""
+    import inspect
```
Why is this `import` here? (rather than globally)

jnoortheen replied:
inspect is heavy to import and add_args will not get called until command is invoked.

daniel-shimon replied:
I see inspect is imported globally throughout the project so it won't add overhead to import it globally here too

jnoortheen replied:
Ok will do.

jnoortheen replied:
updated in the PR

daniel-shimon commented on xonsh/cli_utils.py:
```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
+        if hasattr(annot, "__metadata__"):
+            args, kwargs = annot.__metadata__[0]
+        else:
+            from typing_extensions import get_args
```
This will fail if `typing_extensions` is not installed and `typing.Annotated` is avaliable

jnoortheen replied:
I will check it.

jnoortheen replied:
actually the __metadata__ attribute is available in `typing.Annotated` of py3.9 
the class Annotated is loosely based on that thing. 
for polyfilling previous version one already needs typing_extensions.

daniel-shimon replied:
I see.
So for clarity (and like we do in other files in xonsh), use `get_args` from typing if its available or from `typing_extensions` when it isn't.
Also try importing instead of checking hasattr(tp, "Annotated"). That way we can be sure both `Annotated` and `get_args` exist.

```python
try:
    from typing import Annotated, get_args  # noqa

    TYPING_ANNOTATED_AVAILABLE = True
except ImportError:
    try:
        from typing_extensions import Annotated, get_args  # type: ignore

        TYPING_ANNOTATED_AVAILABLE = True
    except ImportError:
         ...
```

jnoortheen replied:
get_args is used only when typing_extensions is used. we don't have to import that from `typing`.

>  try importing instead of checking hasattr(tp, "Annotated").

I don't have any opinion on this. I am updating.

jnoortheen replied:
Also, get_args is py3.8+ I think.

daniel-shimon replied:
You didn't update this 😕

jnoortheen replied:
Actually I have removed the `if` statement hasattr check and used `importerror`. Here it relies on the fact that Annotated from stdlib has __metadata__ on it.

jnoortheen replied:
@daniel-shimon I've tested without typing-extension on Python3.9 . It doesn't raise any errors. Let me know if there is any other issues.

daniel-shimon commented on xonsh/cli_utils.py:
```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
+    args, kwargs = [], {}
+    if isinstance(annot, tuple):
+        args, kwargs = annot
+    elif TYPING_ANNOTATED_AVAILABLE and "Annotated[" in str(annot):
```
Isn't there a better way than checking the string?

jnoortheen replied:
the other way involves more code and digging some implementation detail. the generics changed in py3.7 from how it used to have __origin__ attrs ...

daniel-shimon commented on xonsh/cli_utils.py:
```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
+        action = parser.add_argument(*args, **kwargs)
+        if completer:
+            action.completer = completer  # type: ignore
+        action.help = action.help or ""
+        if action.default and "%(default)s" not in action.help:
+            action.help += os.linesep + " (default: %(default)s)"
+        if action.type and "%(type)s" not in action.help:
+            action.help += " (type: %(type)s)"
```
After all of this we should set `action.description = action.help.split("\n", 1)[0]` and use it later in `RichCompletion.description` for two reasons:
1. The `%(default)` and `%(type)` don't get formatted in the completion's description
2. The newline doesn't render well as the completion's description

jnoortheen replied:
they are displayed in --help output though. In future, I want to have a completer that can show multi-line description, much like IDEs.

daniel-shimon replied:
That's why I said - keep the help string as is for `--help` and only change the description for RichCompletion.
The multi-line description thing could be cool and can be in a separate pr when you get to it, but for now it doesn't work so let's keep the first line only

jnoortheen replied:
here this is help output. we don't need to truncate here I suppose.

daniel-shimon replied:
Instead of using the same text for both `--help` and `RichCompletion.description` , create a shortened version for the description without these `(default: ...)` and `(type: ...)` fields and only the first line (until newlines can be rendered in ptk)

jnoortheen replied:
The completer should not be making assumptions about the interface. What if I want to use fzf for completions while ptk completion is disabled? It supports multiple lines.

daniel-shimon replied:
Ok that makes sense.
So we need to fill in the `type` and `default` fields when setting `RichCompletion.description`.
Also maybe replace the newlines with a period and a space in `xonsh/ptk` for the time being so we won't have the ugly `^J`?

jnoortheen replied:
still replacing \n should be done inside ptk-completer as it is interface specific. I am not sure about the short representation. don't we want to see if there is a default and what type of argument expected. if there is space to show that info, don't we want to see ?

jnoortheen replied:
created PR for this https://github.com/xonsh/xonsh/pull/4419

daniel-shimon replied:
> don't we want to see if there is a default and what type of argument expected. if there is space to show that info, don't we want to see ?

Yeah I agree! Just last time I checked the type and default didn't work in the description

jnoortheen replied:
![image](https://user-images.githubusercontent.com/6702219/130031973-c02aed3a-5833-41a6-930e-9e1796e9331a.png)

are you talking about --help ?

daniel-shimon replied:
I saw this happen before you split the PRs:
If I change `pos` to `--pos` and add a `%(default)` to the description it doesn't render the value:
![image](https://user-images.githubusercontent.com/18242949/130503905-8f435fa0-0d26-4cde-b640-be5b50e84cf0.png)

jnoortheen replied:
with this https://github.com/xonsh/xonsh/pull/4391/commits/f91e204545892e23fc0f7da18873bc3b89170341 above will get expanded if available.

daniel-shimon commented on xonsh/cli_utils.py:
```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
+            self.commands = self.add_subparsers(title="commands", dest="command")
+
+        doc = get_doc(func)
+        kwargs.setdefault("description", doc)
```
Also here set description to be `help`'s first line (bc the newline doesn't render well in the completion's description)

jnoortheen replied:
this is the content shown in the help as well. I think the responsibility of what to show in the completion-widget should be left to widget itself.

daniel-shimon replied:
Sure, but if it doesn't define one, why include the newlines that we know aren't being displayed correctly currently on ptk?

jnoortheen replied:
Here we don't have to assume it is ptk. Perhaps the ptk completer class can update this completions?

daniel-shimon commented on xonsh/xonfig.py:
```diff
@@ -617,16 +648,28 @@ def _tok_colors(cmap, cols):
+
+
+def _colors(style: Annotated[str, Arg(nargs="?", completer=ColorCompleter())] = None):
+    """preview color style
```
I see some of the descriptions start with a lower case and some with an uppercase. Can you normalize them all to start with an uppercase?

jnoortheen replied:
updated the completer sub-command docstrings. I will put this out when creating PR for xonfig

daniel-shimon commented on xonsh/xonfig.py:
```diff
@@ -617,16 +648,28 @@ def _tok_colors(cmap, cols):
+    Parameters
+    ----------
+    style
+        style to preview, default: <current>
```
Is something supposed to happen with `<current>`?

daniel-shimon commented on xonsh/xontribs.py:
```diff
@@ -78,8 +77,34 @@ def update_context(name, ctx=None):
 
-def xontribs_load(names, verbose=False):
-    """Load xontribs from a list of names"""
+class XontribCompleter(ArgCompleter):
```
Maybe only complete non-loaded xontribs?

jnoortheen replied:
yep a good idea.

jnoortheen replied:
I will be creating separate PRs for all these builtins aliases, once the base class is merged. I will take into account all the other comments in them.

jnoortheen replied:
done locally. collapsing to remind myself

daniel-shimon commented on xontrib/vox.py:
```diff
@@ -75,121 +63,113 @@ def parser():
-        """Create a virtual environment in $VIRTUALENV_HOME with python3's ``venv``."""
+    def new(
+        self,
+        name: Annotated[str, Arg(metavar="ENV")],
```
Writing `Annotated[type, Arg(...)`  for each arg seems like a lot of boilerplate...
Can it be switched to `Arg(type, ...)` that returns an `Annotated[...]` object? I think mypy will work with that
(another possible style can be `Arg[type](...)`)

jnoortheen replied:
1. Arg[] will work here, but we will lose ability to use keyword arguments then
2. I tried Arg() first, with returning type if type_checking is true, but mypy objects anything not in format of `Type[]`

daniel-shimon commented on xontrib/vox.py:
```diff
@@ -75,121 +63,113 @@ def parser():
+        ----------
+        name : str
+            Virtual environment name
+        interpreter: str
+            Python interpreter used to create the virtual environment.
+            Can be configured via the $VOX_DEFAULT_INTERPRETER environment variable.
```
For some reason I don't see `--interpreter`'s description in the tab completion :(

jnoortheen replied:
fixed it. I will update it a new PR. just waiting for #4379

daniel-shimon replied:
Hi is this fixed in #4437?

jnoortheen replied:
Yes. Please make sure to run the aliases locally though.

daniel-shimon commented on xonsh/cli_utils.py:
```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
-        kwargs[name] = ns[name]
+    for name, param in sign.parameters.items():
+        default = None
+        if param.default != inspect.Parameter.empty:
+            default = param.default
```
Isn't this already done in `add_args`?

jnoortheen replied:
yes you are right.

daniel-shimon replied:
I see, can you add a comment explaining this?

daniel-shimon replied:
ping :)

daniel-shimon replied:
ping :)

jnoortheen replied:
done https://github.com/xonsh/xonsh/pull/4391/commits/7ca1cf2d5f0d97cb546b4c21d2d2cb8cf178db50

daniel-shimon commented on xonsh/cli_utils.py:
```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
+        For usage please check ``xonsh.completers.completer.py`` module.
+    """
+
+    def __init__(self, threadable=True, **kwargs) -> None:
```
How should this be used with simple one-function aliases?
I reckon a simple `@arg_parser_alias(threadable=...)` decorator can be useful (e.g. the `xoreutils` aliases)

jnoortheen replied:
![image](https://user-images.githubusercontent.com/6702219/126885236-6849beff-dce0-4757-8f5d-3b3c18e3a3e3.png)

this is how single functions can be added.

daniel-shimon reviewed (commented):
Hi @jnoortheen!
This PR is super awesome and can really improve the xonsh experience!! 🔥 
Sorry there's a ton of comments but there's a lot of ground to cover, hopefully you'll find time to go through all of it 😄

jnoortheen commented on xonsh/completers/completer.py:
```diff
@@ -125,3 +190,41 @@ def remove_completer(name: str):
+    cmd = command.args[0].value
+
+    alias = XSH.aliases.get(cmd)  # type: ignore
+    if not hasattr(alias, "parser"):
```
previously I had that version. But the isinstance thing failed, because of different version from amalgamated class instance and normal import paths. This way one can just set the attribute as well.

daniel-shimon replied:
I see. Can you add a todo comment above? I intend on fixing this amalgamation issue one day

jnoortheen replied:
sure will do. I plan to create a new PR just for ArgParserAlias and completer alias though. I will put the changes there. So you will not be having the noise from changes for other aliases

jnoortheen commented on xonsh/completers/completer.py:
```diff
@@ -125,3 +190,41 @@ def remove_completer(name: str):
+
+
+@contextual_command_completer
+def complete_argparser_aliases(command: CommandContext):
```
I decided to break this PR. in the upcoming PR I will keep this inside completer/_alias.py . I just wanted to merge all things related to completer alias in one place. What do you think about it? which will be better, I don't like the preceding underscore. It has some public functions, like the ones to remove/add completers.

daniel-shimon replied:
I'm fine with creating a new `aliases.py` file for this function, since it **is** the completer for aliases :)
I rather you do it in this PR if you don't mind, generally I intend on releasing versions more frequently and I prefer inserting it correctly in the first place rather than inserting and then refactoring

daniel-shimon replied:
I'm continuing the threads here instead of re-opening them in https://github.com/xonsh/xonsh/pull/4391

daniel-shimon replied:
Also `ArgparseCompleter` should be there

jnoortheen replied:
> Also `ArgparseCompleter` should be there

it is now in the `completer/tools.py` along other completion helpers.

daniel-shimon replied:
It isn't a tool for other completera to use, it's the completion logic for aliases. 
Just like the bash completion logic is in `bash.py`, and completion logic for paths is in `paths.py`, the completion logic for aliases needs to be in an `aliases.py` file.
So goes for all the other logic you put in `completer.py`

jnoortheen replied:
moved `complete_argparser_aliases` to completers/_aliases.py  in 
 
https://github.com/xonsh/xonsh/pull/4391/commits/2e45190436ebdd52e874f28142e3dbea20db05a3

daniel-shimon replied:
Same for `ArgparseCompleter` and `complete_argparser`
Sorry I'm still pressing this it's just that this bloats `completers/tools.py` with things that are specific to the aliases completer only

jnoortheen replied:
the completion logic is for argparse.Parser. For some external command, one can build a completer from this class. This is not specific to alias.

jnoortheen replied:
anyways I will move this class to `cli_utils.py` if you are concerned.

jnoortheen replied:
I have created its own module inside completers.

jnoortheen replied:
@daniel-shimon I think this is the final thing to resolve. Do I need to update the amalgamate setting now?

daniel-shimon replied:
This can be resolved now :)
Go ahead!

jnoortheen commented on xonsh/completers/completer.py:
```diff
@@ -125,3 +190,41 @@ def remove_completer(name: str):
+    possible = complete_argparser(alias.parser, command=command, alias=alias)
+    for comp in possible:
+        if comp.startswith(command.prefix):
+            yield comp
```
Yes that will be easier to implement. This version still has a todo left :)
I want to see all the options available for a command in one glance. that way I can decide on arguments that I want to use.

daniel-shimon replied:
The standard (in xonsh and other shells) is to only show these once the user has wrote the first dash, it's what users expect and I don't see any reason to do otherwise

jnoortheen replied:
added an environment variable to control this behaviour

https://github.com/xonsh/xonsh/pull/4391/commits/db5a4f88bb624e1694429e8c80fae3ac5324afe6

jnoortheen commented on xonsh/xonfig.py:
```diff
@@ -617,16 +648,28 @@ def _tok_colors(cmap, cols):
 
 
-def _colors(args):
+class ColorCompleter(ArgCompleter):
+    def __call__(self, *args, **kwargs):
```
yes it supports passing a function as completer. It checks for only callables. 
Class is used to provide some structure and extensibility. e.g. we can have completer for path as basic with attributes for skipping some, having globs ....

daniel-shimon replied:
Sounds good, but still for all completers in this PR I don't see any reason to add the boilerplate.. You can enforce the structure via type hints in `Arg`

jnoortheen replied:
Ok. I will refactor them to simple functions.

jnoortheen commented on xonsh/completers/completer.py:
```diff
@@ -1,51 +1,110 @@
+
+class CompletersChoices(ArgCompleter):
+    def __call__(self, xsh, **_):
+        yield from xsh.completers.keys()
```
I am not sure what to show in the description. do you have any suggestion?

jnoortheen replied:
Oh got it thanks

jnoortheen commented on xonsh/completers/completer.py:
```diff
@@ -125,3 +190,41 @@ def remove_completer(name: str):
+
+    possible = complete_argparser(alias.parser, command=command, alias=alias)
+    for comp in possible:
+        if comp.startswith(command.prefix):
```
done.

jnoortheen commented on xonsh/xonfig.py:
```diff
@@ -715,87 +782,25 @@ def _jupyter_kernel(args):
+        parser = self.create_parser(prog="xonfig")
+        # register as default action
+        add_args(parser, _info, allowed_params=())
+        parser.add_command(_info, prog="info")
```
these function names are generic. after amalgamation, it may collide with other names.

jnoortheen commented:
@daniel-shimon are there any other things to check?

anki-code commented:
The coloring is working for me:
![image](https://user-images.githubusercontent.com/1708680/131486782-daf69833-3ca0-41fa-b32a-afa7c5ffb83a.png)

jnoortheen commented:
@anki-code this PR is outdated. I will close it after the related PRs get closed. 

please check the one in here https://github.com/xonsh/xonsh/pull/4437
